File: Adam_FastQ_SequenceQual_Baseline/src/main/scala/FastQ_PerSequenceQual_Job.scala
package main.scala

import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.io.Text
import org.apache.hadoop.mapreduce.lib.input.FileSplit
import org.apache.parquet.hadoop.util.ContextUtil
import org.apache.spark.rdd.{NewHadoopRDD, RDD}
import org.apache.spark.{SparkConf, SparkContext}
import org.bdgenomics.adam.io.SingleFastqInputFormat
import org.bdgenomics.formats.avro.AlignmentRecord
import org.bdgenomics.utils.misc.HadoopUtil

/**
  * master-thesis Clemens Banas
  * Organization: DBIS - University of Innsbruck
  * Created 05.02.16.
  */
object FastQ_PerSequenceQual_Job {

  def main(args: Array[String]): Unit = {
    if (args.length != 2) {
      println("usage: spark-submit Adam_FastQ_SequenceQual_Baseline-1.0-SNAPSHOT.jar <fastq input directory> <output dir>")
      return;
    }

    val input = args(0)
    val output = args(1)

    val conf = new SparkConf()
    conf.setAppName("Adam_FastQ_SequenceQual_Baseline")
    val sc = new SparkContext(conf)

    val configuration = new Configuration()
    configuration.set("mapreduce.input.fileinputformat.inputdir", input)

    val job = HadoopUtil.newJob(sc)
    val fastQInput = sc.newAPIHadoopFile(
      input,
      classOf[SingleFastqInputFormat],
      classOf[Void],
      classOf[Text],
      ContextUtil.getConfiguration(job)
    )

    val hadoopRdd = fastQInput.asInstanceOf[NewHadoopRDD[Void, Text]]
    val myRdd: RDD[Pair[String, Text]] = hadoopRdd.mapPartitionsWithInputSplit { (inputSplit, iterator) ⇒
      val file = inputSplit.asInstanceOf[FileSplit]
      iterator.map(record => (file.getPath.getName, record._2))
    }

    val adamRDD: RDD[Pair[String, AlignmentRecord]] = myRdd.map(record => FastQ_PerSequenceQual_Mapper.mapSampleIdentifierWithConvertedInputObject(record))

    //mapping step
    val qualityScores: RDD[Pair[Pair[String, Int], Int]] = adamRDD.map(record => FastQ_PerSequenceQual_Mapper.map(record._1, record._2))

    //reduce step
    val res = qualityScores.reduceByKey(
      (a: Int, b: Int) => FastQ_PerSequenceQual_Reducer.combine(a, b)
    )

    //format and save output to file
    res.map(record => record._1._1 + "," + record._1._2 + "," + record._2).saveAsTextFile(output)
  }

}

File: MapReduce_FastQ_SequenceQual_Combiner/src/main/java/sequenceQual/FastQ_PerSequenceQual_Job.java
package sequenceQual;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
import org.seqdoop.hadoop_bam.FastqInputFormat;
import utils.IdentifierPositionKeyWritable;

/**
 * master-thesis Clemens Banas
 * Organization: DBIS - University of Innsbruck
 * Created 21.04.2016
 */

public class FastQ_PerSequenceQual_Job {

    public static void main(String[] args) throws Exception {
        Configuration conf = new Configuration();
        Job job = Job.getInstance(conf, "MR_FastQ_PerSequenceQual_Combiner");

        job.setJarByClass(FastQ_PerSequenceQual_Job.class);
        job.setMapperClass(FastQ_PerSequenceQual_Mapper.class);
        job.setCombinerClass(FastQ_PerSequenceQual_Combiner.class);
        job.setReducerClass(FastQ_PerSequenceQual_Reducer.class);

        job.setInputFormatClass(FastqInputFormat.class);
        job.setMapOutputKeyClass(IdentifierPositionKeyWritable.class);
        job.setMapOutputValueClass(IntWritable.class);

        job.setOutputKeyClass(IdentifierPositionKeyWritable.class);
        job.setOutputValueClass(IntWritable.class);

        job.setOutputFormatClass(FastQ_PerSequenceQual_OutputFormat.class);

        FileInputFormat.addInputPath(job, new Path(args[0]));
        FileOutputFormat.setOutputPath(job, new Path(args[1]));

        if (!job.waitForCompletion(true)) {
            System.err.println("sort :: Job failed.");
        }
    }

}
